L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleGen.cc

```diff
@@ -40,53 +205,245 @@ HGCalTriggerNtupleGen(const edm::ParameterSet& conf):HGCalTriggerNtupleBase(conf
     const std::vector< PileupSummaryInfo >& PupInfo = *PupInfo_h;
 
-    clear();
-    gen_n_ = gen_particles.size();
-    gen_id_.reserve(gen_n_);
-    gen_status_.reserve(gen_n_);
-    gen_energy_.reserve(gen_n_);
-    gen_pt_.reserve(gen_n_);
-    gen_eta_.reserve(gen_n_);
-    gen_phi_.reserve(gen_n_);
-    for(const auto& particle : gen_particles)
-    {
-        gen_id_.emplace_back(particle.pdgId());
-        gen_status_.emplace_back(particle.status());
-        gen_energy_.emplace_back(particle.energy());
-        gen_pt_.emplace_back(particle.pt());
-        gen_eta_.emplace_back(particle.eta());
-        gen_phi_.emplace_back(particle.phi());
+
+    // FIXME: this part could go in begin run
+    edm::ESHandle<HepPDT::ParticleDataTable> pdt;
+    es.getData(pdt);
+    mySimEvent_->initializePdt(&(*pdt));
+
+    retrieveLayerPositions(es, 52);
+
+    edm::ESHandle<MagneticField> magfield;
+    es.get<IdealMagneticFieldRecord>().get(magfield);
+    aField_ = &(*magfield);
+    // up to here...could go in the beginRun
+
+    // This balck magic is needed to use the mySimEvent_
+    ParticleTable::Sentry ptable(mySimEvent_->theTable());
+    edm::Handle<edm::HepMCProduct> hevH;
+    edm::Handle<std::vector<SimTrack>> simTracksHandle;
+    edm::Handle<std::vector<SimVertex>> simVerticesHandle;
+
+    iEvent.getByToken(hev_token_, hevH);
+    iEvent.getByToken(simTracks_token_, simTracksHandle);
+    iEvent.getByToken(simVertices_token_, simVerticesHandle);
+    mySimEvent_->fill(*simTracksHandle, *simVerticesHandle);
+
+    HepMC::GenVertex *primaryVertex = *(hevH)->GetEvent()->vertices_begin();
+    vtx_x_ = primaryVertex->position().x() / 10.;  // to put in official units
+    vtx_y_ = primaryVertex->position().y() / 10.;
+    vtx_z_ = primaryVertex->position().z() / 10.;
+    Point sim_pv(vtx_x_, vtx_y_, vtx_z_);
+
+
+    HGCal_helpers::simpleTrackPropagator toHGCalPropagator(aField_);
+    toHGCalPropagator.setPropagationTargetZ(layerPositions_[0]);
+    std::vector<FSimTrack *> allselectedgentracks;
+    unsigned int npart = mySimEvent_->nTracks();
+    for (unsigned int i = 0; i < npart; ++i) {
+      std::vector<float> xp, yp, zp;
+      FSimTrack &myTrack(mySimEvent_->track(i));
+      math::XYZTLorentzVectorD vtx(0, 0, 0, 0);
+
+      int reachedEE = 0;  // compute the extrapolations for the particles reaching EE
+                          // and for the gen particles
+      double fbrem = -1;
+
+      if (std::abs(myTrack.vertex().position().z()) >= layerPositions_[0]) continue;
+
+      unsigned nlayers = 40;
+      if (myTrack.noEndVertex())  // || myTrack.genpartIndex()>=0)
+      {
+        HGCal_helpers::coordinates propcoords;
+        bool reachesHGCal = toHGCalPropagator.propagate(
+            myTrack.momentum(), myTrack.vertex().position(), myTrack.charge(), propcoords);
+        vtx = propcoords.toVector();
+
+        if (reachesHGCal && vtx.Rho() < 160 && vtx.Rho() > 25) {
```

Comment:
Avoid magic numbers `160` and `25`

Reply:
Ideally these would not be hardcoded but I wouldn't know where to pull these numbers from.
Ideas are welcome.

Reply:
It can be hardcoded, but a meaningful variable name is still better than just raw numbers.
These are the inner and outer radius of HGCal, so you could have:
```cpp
const float hgcalInnerRadius = 25.;
const float hgcalOuterRadius = 160.;
if (reachesHGCal && vtx.Rho() < hgcalOuterRadius && vtx.Rho() > hgcalInnerRadius) {
```